neon_llm_chatgpt/chatgpt.py

```diff
@@ -25,39 +25,132 @@
+            Answers are sorted from best to worst
+            :param question: incoming question
+            :param answers: list of answers to rank
+            :returns list of indexes
+        """
+        if not answers:
+            return []
+        scores = self._score(prompt=question, targets=answers)
+        sorted_items = sorted(zip(range(len(answers)), scores), key=lambda x: x[1])
+        sorted_items_indexes = [x[0] for x in sorted_items]
+        return sorted_items_indexes
+
+    def _call_model(self, prompt: List[Dict[str, str]]) -> str:
+        """
+            Wrapper for ChatGPT Model generation logic
+            :param prompt: Input messages sequence
+            :returns: Output text sequence generated by model
+        """
+
+        response = openai.ChatCompletion.create(
+            model=self.llm_model_name,
+            messages=prompt,
+            temperature=0,
+            max_tokens=self.max_tokens,
+        )
+        text = response.choices[0].message['content']
+
+        return text
+
+    def _assemble_prompt(self, message: str, chat_history: List[List[str]]) -> List[Dict[str, str]]:
+        """
+            Assembles prompt engineering logic
+            Setup Guidance:
+            https://platform.openai.com/docs/guides/gpt/chat-completions-api
+
+            :param message: Incoming prompt
+            :param chat_history: History of preceding conversation
+            :returns: assembled prompt
+        """
         messages = [
-            {"role": "system", "content": self.role},
+            {"role": "system", "content": self._system_prompt},
         ]
         # Context N messages
         for role, content in chat_history[-self.context_depth:]:
             role_chatgpt = self.convert_role(role)
             messages.append({"role": role_chatgpt, "content": content})
         messages.append({"role": "user", "content": message})
-        
-        response = openai.ChatCompletion.create(
-            model=self.model,
-            messages=messages,
-            temperature=0,
-            max_tokens=self.max_tokens,
-        )
-        bot_message = response.choices[0].message['content']
-        return bot_message
+        return messages
+
+    def _score(self, prompt: str, targets: List[str]) -> List[float]:
+        """
+            Calculates logarithmic probabilities for the list of provided text sequences
+            :param prompt: Input text sequence
+            :param targets: Output text sequences
+            :returns: List of calculated logarithmic probabilities per output text sequence
+        """
+
+        question_embeddings, answers_embeddings = self._embeddings(question=prompt, answers=targets)
+        scores_list = distances_from_embeddings(question_embeddings, answers_embeddings)
+        return scores_list
+
+    def _tokenize(self, prompt: str) -> None:
+        return None
```

Comment:
better use `pass`

Reply:
Is this a `#TODO`, or is this intentionally left empty?

Reply:
@NeonDaniel intentionally, or else there will be an error, because it is defined in parent abstract class
But chatgpt handles tokenization internally

Reply:
Done with https://github.com/NeonGeckoCom/neon-llm-chatgpt/pull/9/commits/6b899779cd5572f07a83fa404541654e6900e2fd